### PR TITLE
[ONNX] Implement symbolic dispatcher

### LINKED ISSUE #82046: [ONNX] Op version agnostic symbolics
Currently we go back and forth from versioned symbolic functions to symbolic_helper that condition on the current opset version. A dispatching mechanism for separating versioning decisions from the symbolic mapping would make the symbolic functions easier to interpret.

Related: https://github.com/pytorch/pytorch/issues/78391

Reference: tf2onnx

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85488
* #85487
* __->__ #85486
* #85350
* #85349

Dispatcher for #83787 . A pyi file will be created for external use in a separate PR.

### CODE REVIEW

File: torch/onnx/_internal/dispatch/dispatch.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    context = args[0]
+    assert isinstance(context, torchscript.GraphContext)
+    symbolic_fn = registration.registry.get_function_group(qualified_name)
+    if symbolic_fn is None:
+        raise errors.UnsupportedOperatorError(
+            qualified_name,
+            context.opset,
+            None,
+        )
+    return symbolic_fn(context, *args, **kwargs)
```
Comment (justinchuby) [inline torch/onnx/_internal/dispatch/dispatch.py:61]: This is still a two step resolution done in different places. I wonder if there is a clearer way of doing it?

**Comment by justinchuby** [comment]
cc @thiagocrepaldi for comments

**Comment by thiagocrepaldi** [comment]
Hi Justin, not having a design or at least a detailed PR description has the implications that we already discussed in our weekly meeting (strengthening the design for the non happy-path, spread knowledge in the team and document the new code), but it also costs code-reviewers a lot of time to guess what is not written besides the source code (the problem to be solved, why, what, when and how), possible accidental limitations, dig both old and new code to have more context, etc

My understanding after the last meeting in which the need-for-design was proposed was that non bug-fix changes should be blocked until a design was proposed/discussed.

I will wait on @abock to confirm my understanding and/or approve this as an exception

**Comment by justinchuby** [comment]
@thiagocrepaldi thanks for the suggestion. You are absolutely right. More details will be filled in on https://github.com/pytorch/pytorch/issues/82046 (currently just a motivation). I was hoping to get your initial feedback for the prototypes. Feel free to wait for more context if that's what you prefer too.

**Pull request closed without merging**